fft/src/KokkosFFT_ROCM_types.hpp

```diff
@@ -17,27 +25,227 @@ static_assert(sizeof(std::complex<double>) == sizeof(Kokkos::complex<double>));
+  rocfft_plan m_plan;
+  rocfft_execution_info m_execution_info;
+
+  bool m_is_plan_created = false;
+  bool m_is_info_created = false;
+
+  //! Internal work buffer
+  BufferViewType m_buffer;
+
+ public:
+  ScopedRocfftPlan(const Kokkos::HIP &exec_space,
+                   const FFTWTransformType transform_type,
+                   const std::vector<int> &in_extents,
+                   const std::vector<int> &out_extents,
+                   const std::vector<int> &fft_extents, int howmany,
+                   Direction direction, bool is_inplace) {
+    auto [in_array_type, out_array_type, fft_direction] =
+        get_in_out_array_type(transform_type, direction);
+
+    // Compute dist and strides from extents
+    int idist = std::accumulate(in_extents.begin(), in_extents.end(), 1,
+                                std::multiplies<>());
+    int odist = std::accumulate(out_extents.begin(), out_extents.end(), 1,
+                                std::multiplies<>());
+
+    auto in_strides  = compute_strides<int, std::size_t>(in_extents);
+    auto out_strides = compute_strides<int, std::size_t>(out_extents);
+    auto reversed_fft_extents =
+        convert_int_type_and_reverse<int, std::size_t>(fft_extents);
+
+    // Create a plan description
+    ScopedRocfftPlanDescription scoped_description;
+    rocfft_status status = rocfft_plan_description_set_data_layout(
+        scoped_description.description(),  // description handle
+        in_array_type,                     // input array type
+        out_array_type,                    // output array type
+        nullptr,                           // offsets to start of input data
+        nullptr,                           // offsets to start of output data
+        in_strides.size(),                 // input stride length
+        in_strides.data(),                 // input stride data
+        idist,                             // input batch distance
+        out_strides.size(),                // output stride length
+        out_strides.data(),                // output stride data
+        odist);                            // output batch distance
+
+    KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                       "rocfft_plan_description_set_data_layout failed");
+
+    // inplace or Out-of-place transform
+    const rocfft_result_placement place =
+        is_inplace ? rocfft_placement_inplace : rocfft_placement_notinplace;
+
+    // Create a plan
+    status = rocfft_plan_create(&m_plan, place, fft_direction, m_precision,
+                                reversed_fft_extents.size(),  // Dimension
+                                reversed_fft_extents.data(),  // Lengths
+                                howmany,  // Number of transforms
+                                scoped_description.description()  // Description
+    );
+    KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                       "rocfft_plan_create failed");
```

Comment:
I think we should try to make the last thing to call in the constructor, if there is any exception throw after the plan creation, the destructor will never be called and the plan will never to freed.

Reply:
Could you please detail this point?
If there is an exception after plan creation, `m_plan` is supposed to be destroyed in destructor of `ScopedRocfftPlan`

Reply:
Maybe there was confusion, by plan creation I meant the creation of `m_plan`.